app/controllers/task_contributions_controller.rb

```diff
@@ -37,42 +37,51 @@ def show
   end
 
   def new
-    @task = Task.find(params[:task_id]).clean_duplicate(current_user, change_title: false)
-    @task.parent_id = params[:task_id]
-    task_contrib = TaskContribution.new
-    @task.task_contribution = task_contrib
-    authorize task_contrib
-    render 'new'
+    @task_contribution = TaskContribution.new_for(@task, current_user)
+    authorize @task_contribution
+
+    @task = @task_contribution.suggestion
+    render 'tasks/new'
   end
 
   # The function should render the edit form used by TaskController
   def edit
-    contrib = TaskContribution.find(params[:id])
-    @task = contrib.modifying_task
+    @task = @task_contribution.suggestion
     render 'tasks/edit'
   end
 
   def create
-    @task = Task.new(contrib_task_params)
-    contrib = TaskContribution.new(status: :pending)
-    @task.task_contribution = contrib
-    authorize contrib
-    if @task.save(context: :force_validations)
-      TaskContributionMailer.contribution_request(contrib).deliver_later
-      redirect_to @task, notice: t('.success')
+    @task_contribution = TaskContribution.new(suggestion_attributes: task_params, base: @task)
+    @task_contribution.suggestion.assign_attributes(user: current_user, access_level: :private)
+    authorize @task_contribution
+
+    if @task_contribution.save(context: :force_validations)
+      TaskContributionMailer.contribution_request(@task_contribution).deliver_later
+      redirect_to [@task, @task_contribution], notice: t('.success')
     else
-      redirect_to Task.find(params[:task_id]), alert: t('.error')
+      redirect_to @task, alert: t('.error')
     end
   end
 
-  def contrib_task_params
-    task_params
-      .merge(user: current_user, parent_uuid: Task.find(params[:task_id]).uuid,
-        access_level: :private, task_contribution: TaskContribution.new)
+  def update
+    @task_contribution.suggestion.assign_attributes(task_params.except(:parent_uuid))
+    if @task_contribution.save(context: :force_validations)
+      redirect_to [@task, @task_contribution], notice: t('.success')
+    else
+      @task = @task_contribution.suggestion
+      render 'tasks/edit'
+    end
+  end
+
+  private
+
+  def load_and_authorize_task
+    @task = Task.find(params[:task_id])
+    authorize @task, :show?
```

Comment:
Potentially, we could move it. Still, there is one difference: The authorization action is different. In the `TaskController`, we are just using the general `authorize @task`, which will use the current action for authorization. Here, in the `TaskContributionsController`, we are always checking the `show?` action for `@task`, since the _actual_ object and item to be checked is the `@task_contribution`. 

Potentially, one could work around this by checking the current controller name in the concern (using [`controller_name`](https://api.rubyonrails.org/v7.1.3/classes/ActionController/Metal.html#method-c-controller_name)), which would make the method a bit more "complex":

```rb
  def load_and_authorize_task
    @task = Task.find(params[:task_id])
    if controller_name == 'task_contributions'
      authorize @task, :show?
    else
      authorize @task
    end
  end
```

Reply:
It appears to create more code than anticipated.

I think that your second suggestion looks pretty good but I am not sure if it has any benefits other your original implementation.

What is the general style in similar situations?

Reply:
Thanks for your feedback. Indeed, if we would like to proceed here, the second suggestion (with the dedicated methods in all three files) would probably be the way to go. Comparing to the "original" implementation currently in this PR, the (second) suggestion compares as follows:

+ ➕ The existing code is not mixing the parameter and authorization in different controllers. Especially for a security aspect that helps to reduce the back-and-forward jumps when following the code flow.
- ➖ The existing code looks very similar besides these two minor changes (param name + authorize action). It is not clearly visible (without this discussion) why they are not sharing any code.
+ ➕ The existing code is more similar with all other controllers, having the same structure. We are not deviating for two controllers from the given architecture.
- ➖ The suggestion splits the code into smaller aspects with clearer effects. There, we are introducing two dedicated functions with single responsibilities. Potentially, this system could be generalised even further and reused for other controllers (However, we would then need to define even more smaller helper methods).
- ➕  The existing code is shorter, overall.

My summary would be to stay with the current implementation, rather than using the suggestion. While, yes, there is some code duplication, the alternative is even longer and distributing responsibilities between various files 🍝. I would rather go with _convention_ (same name, same structure, etc.) here rather than with code sharing.